homeassistant/components/wiffi/__init__.py

```diff
@@ -0,0 +1,166 @@
+"""Component for wiffi support."""
+import asyncio
+from datetime import timedelta
+import errno
+import logging
+
+from wiffi import WiffiTcpServer
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PORT
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry
+from homeassistant.helpers.event import async_track_time_interval
+
+from .binary_sensor import BoolEntity
+from .const import DOMAIN
+from .sensor import NumberEntity, StringEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+PLATFORMS = ["sensor", "binary_sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the wiffi component. config contains data from configuration.yaml."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Set up wiffi from a config entry, config_entry contains data from config entry database."""
+    # create api object
+    api = WiffiIntegrationApi(hass, config_entry)
+
+    # store api object
+    hass.data.setdefault(DOMAIN, {})[config_entry.entry_id] = api
+
+    try:
+        await api.server.start_server()
+    except OSError as exc:
+        if exc.errno != errno.EADDRINUSE:
+            raise
+        _LOGGER.error("port %s already in use", config_entry.data[CONF_PORT])
+        raise ConfigEntryNotReady from exc
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Unload a config entry."""
+    api = hass.data[DOMAIN][config_entry.entry_id]  # type ::= WiffiIntegrationApi
+    await api.server.close_server()
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        api = hass.data[DOMAIN].pop(config_entry.entry_id)
+        api.shutdown()
+
+    return unload_ok
+
+
+class WiffiIntegrationApi:
+    """API object for wiffi handling. Stored in hass.data."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the instance."""
+        self._hass = hass
+        self._server = WiffiTcpServer(config_entry.data[CONF_PORT], self)
+        self._known_devices = {}
+        self._async_add_entities = {}
+        self._periodic_callback = async_track_time_interval(
```

Comment:
We don't want side effects in init method. Please move it to a dedicated setup method.